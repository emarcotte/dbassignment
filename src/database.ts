type Key = string;
type Value = string;


/**
 * Transactions are used to hold data. They're layered together to implement nested transactions.
 * The transactions attempt to model a "copy-on-write" model.
 */
export class Transaction {
    /**
     * Parent transaction to use when fetching or setting data. Will be null
     * for the root transaction.
     */
    parent: Transaction | null;

    /**
     * `null` in a transaction indicates that this transaction deleted the key.
     */
    data: Record<Key, Value | null>;

    /**
     * An "index" of the values from the data table. Each value is mapped back to
     * the keys that matched it.
     */
    value_index: Record<Value, Set<Key>>;

    constructor(parent: Transaction | null) {
        this.parent = parent;
        this.data = {};
        this.value_index = {};
    }

    get(key: Key): Value | undefined {
        const value = this.data[key];

        if (value === undefined) {
            if (this.parent == null) {
                return undefined;
            }
            return this.parent.get(key);
        }

        if (value === null) {
            return undefined;
        }
        return value;
    }

    set(key: Key, value: Value): Value {
        const old_value = this.get(key);
        if (old_value != undefined) {
            this.cow_index(old_value).delete(key);
        }

        this.cow_index(value).add(key);
        return this.data[key] = value;
    }

    keys_by_value(search_value: Value): Set<Key> {
        const keys = this.value_index[search_value];
        if (keys) {
            return keys;
        }
        if (this.parent) {
            return this.parent.keys_by_value(search_value);
        }
        return new Set();
    }

    /**
     * Ensures the given index exists on this transaction when we need to write to it.
     */
    cow_index(value: Value): Set<Key> {
        if (this.value_index[value] != undefined) {
            return this.value_index[value];
        }
        if (this.parent) {
            // deeply nested transactions do have 
            return this.value_index[value] = new Set(this.parent.cow_index(value));
        }
        return this.value_index[value] = new Set();
    }

    delete(key: Key): Value | undefined {
        const tx_value = this.data[key];
        const inner_value = this.get(key);

        const index_to_mutate = tx_value || inner_value;

        if (index_to_mutate) {
            this.cow_index(index_to_mutate).delete(key);
        }
        this.data[key] = null;
        return index_to_mutate;
    }

    commit(): Transaction {
        if (!this.parent) {
            return this;
        }

        const parent = this.parent;

        Object.keys(this.value_index).forEach(v => {
            parent.value_index[v] = this.value_index[v];
        });

        Object.keys(this.data).forEach(k => {
            parent.data[k] = this.data[k];
        });

        return parent.commit();
    }
}

/**
 * An extremely "poor-mans" database that is should never see the light
 * of day. This database acts like a simple stringy-key-value store.
 *
 * This database implements the "read my own data" style of transaction where uncommitted
 * transaction state is visible to the client.
 *
 * Note:
 *
 * 1. Nothing is "nullable", absence in the database is expressed with `undefined`.
 * 2. The CLI renders absence of value as "NULL", which is not visual 
 *    distinct from a string with the same value.
 * 3. "committed" data lives in a root transaction, and it s not persisted to disk.
 * 4. Open, uncommitted transactions are layered on top of the root transaction to model
 *    nested transactions. They're merged into their parent on commit.
 */
export class AssignmentDB {
    active_tx: Transaction;

    constructor() {
        this.active_tx = new Transaction(null);
    }

    get(key: Key): Value | undefined {
        return this.active_tx.get(key);
    }

    set(key: Key, value: Value): Value {
        return this.active_tx.set(key, value);
    }

    count(search_value: Key): number {
        return this.active_tx.keys_by_value(search_value).size;
    }

    delete(key: Key): Value | undefined {
        return this.active_tx.delete(key);
    }

    begin(): Transaction {
        this.active_tx = new Transaction(this.active_tx);
        return this.active_tx;
    }

    rollback(): Transaction | null {
        if (this.active_tx.parent) {
            this.active_tx = this.active_tx.parent;
            return this.active_tx;
        }
        return null;
    }

    commit() {
        this.active_tx = this.active_tx.commit();
    }
}
