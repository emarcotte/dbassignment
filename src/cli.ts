import { AssignmentDB } from "./database";
import * as readline from 'node:readline/promises';
import { stdin, stdout } from 'node:process';
import events from 'events';

const command_regex = /^([A-Z]+)(.*)/;
const set_args = /^([^\s]+)\s([^\s]+)$/;
const get_args = /^([^\s]+)$/;
const delete_args = /^([^\s]+)$/;

const handlers: Record<string, (db: AssignmentDB, args: string, output: NodeJS.WritableStream) => boolean> = {
    SET: set_handler,
    GET: get_handler,
    COUNT: count_handler,
    DELETE: delete_handler,
    BEGIN: begin_handler,
    ROLLBACK: rollback_handler,
    COMMIT: commit_handler,
    END: end_handler,
}

function end_handler(): boolean {
    return true;
}

function count_handler(db: AssignmentDB, value: string, output: NodeJS.WritableStream): boolean {
    const count = db.count(value)
    output.write(`${count}\n`);
    return false;
}

function delete_handler(db: AssignmentDB, args: string, output: NodeJS.WritableStream): boolean {
    const match = delete_args.exec(args);
    if (!match) {
        output.write("Invalid DELETE\n");
    }
    else {
        db.delete(match[1]);
    }
    return false;
}

function commit_handler(db: AssignmentDB): boolean {
    db.commit();
    return false;
}

function begin_handler(db: AssignmentDB): boolean {
    db.begin();
    return false;
}

function rollback_handler(db: AssignmentDB, _: string, output: NodeJS.WritableStream): boolean {
    const t = db.rollback();
    if (!t) {
        output.write("TRANSACTION NOT FOUND\n");
    }
    return false;
}

function get_handler(db: AssignmentDB, args: string, output: NodeJS.WritableStream): boolean {
    const match = get_args.exec(args);
    if (!match) {
        output.write("Invalid GET\n");
    }
    else {
        const v = db.get(match[1]);
        if (v === undefined) {
            output.write(`NULL\n`);
        }
        else {
            output.write(`${v}\n`);
        }
    }

    return false;
}

function set_handler(db: AssignmentDB, args: string, output: NodeJS.WritableStream): boolean {
    const match = set_args.exec(args);
    if (!match) {
        output.write("Invalid SET\n");
    }
    else {
        db.set(match[1], match[2]);
    }
    return false;
}

export async function main(
    db: AssignmentDB,
    input: NodeJS.ReadableStream,
    output: NodeJS.WritableStream,
) {
    const ui = readline.createInterface({
        input,
        output,
        prompt: '>> ',
    });
    ui.prompt();

    await Promise.race([
        // If the stdin is closed, see ya later.
        events.once(ui, 'close'),
        // Otherwise, just keep running the line parser
        new Promise<void>((resolve) => {
            ui.on('line', (line) => {
                const match = command_regex.exec(line);
                if (!match) {
                    output.write(`dont know what to do with ${line}\n`);
                }
                else {
                    const args = (match[2] || "").trim();
                    const handler = handlers[match[1]];
                    if (!handler) {
                        output.write(`Invalid command ${match[1]}\n`)
                    }
                    else {
                        if (handler(db, args, output)) {
                            resolve()
                        }
                    }
                }
                ui.prompt();
            });
        }),
    ]);


    ui.close();
    output.end();
}

if (require.main === module) {
    main(new AssignmentDB(), stdin, stdout);
}
