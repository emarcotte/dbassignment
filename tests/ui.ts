import { describe, expect, test } from '@jest/globals';
import { AssignmentDB } from '../src/database';
import { main } from '../src/cli';
import { PassThrough } from 'stream';
import { stdin as mockStdIn, MockSTDIN } from 'mock-stdin';

async function inputs(stdin: MockSTDIN, inputs: string[]): Promise<void> {
    for (const i of inputs) {
        await new Promise(process.nextTick);
        stdin.send(`${i}\n`);
    }
}

async function readAll(stream: PassThrough): Promise<string> {
    let b = Buffer.from([]);
    for await (const chunk of stream) {
        b = Buffer.concat([b, chunk]);
    }
    return b.toString();
}

async function drive_ui(data: string[]): Promise<{ db: AssignmentDB, output: string }> {
    const db = new AssignmentDB();
    const stdin = mockStdIn();
    const stdout = new PassThrough();
    const app = main(db, process.stdin, stdout);
    await inputs(stdin, data);
    await app;
    const output = await readAll(stdout);
    return { db, output };
}

describe("UI", () => {
    test("basic set/get inputs", async () => {
        const { db, output } = await drive_ui([
            "SET 1 1234",
            "GET 1",
            "END",
        ]);

        expect(db.get('1')).toEqual('1234');
        expect(output).toMatchSnapshot();
    });

    test("count basics", async () => {
        const { db, output } = await drive_ui([
            "SET 1 1234",
            "COUNT 1234",
            "COUNT 123",
            "SET 2 1234",
            "COUNT 1234",
            "END",
        ]);

        expect(db.get('1')).toEqual('1234');
        expect(output).toMatchSnapshot();
    });

    test("invalid commands", async () => {
        const { output } = await drive_ui([
            "hello",
            "HELLO",
            "GET",
            // disallow spaces in keys
            "GET a a",
            "SET",
            "SET X",
            // disallow spaces in values (or keys)
            "SET X a a",
            "DELETE",
            "END",
        ]);

        expect(output).toMatchSnapshot();
    });

    test("commit", async () => {
        const { db, output } = await drive_ui([
            "COMMIT",
            "SET 1 1234",
            "BEGIN",
            "GET 1",
            "SET 1 123",
            "BEGIN",
            "SET 1 abc",
            "COMMIT",
            "GET 1",
            "ROLLBACK",
            "END",
        ]);

        expect(db.get('1')).toEqual('abc');
        expect(output).toMatchSnapshot();

    })

    test("basic rollback UI", async () => {
        const { db, output } = await drive_ui([
            "ROLLBACK",
            "SET 1 1234",
            "BEGIN",
            "GET 1",
            "SET 1 123",
            "GET 1",
            "ROLLBACK",
            "GET 1",
            "END",
        ]);

        expect(db.get('1')).toEqual('1234');
        expect(output).toMatchSnapshot();
    });

    test("delete", async () => {
        const { db, output } = await drive_ui([
            "SET 1 1234",
            "GET 1",
            "DELETE 1",
            "GET 1",
            "END",
        ]);

        expect(db.get('1')).toBeUndefined()
        expect(output).toMatchSnapshot();
    });
});
