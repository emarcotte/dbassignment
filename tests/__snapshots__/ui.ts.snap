// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI basic rollback UI 1`] = `
">> TRANSACTION NOT FOUND
>> >> >> 1234
>> >> 123
>> >> 1234
>> >> "
`;

exports[`UI basic set/get inputs 1`] = `
">> >> 1234
>> >> "
`;

exports[`UI commit 1`] = `
">> >> >> >> 1234
>> >> >> >> >> abc
>> TRANSACTION NOT FOUND
>> >> "
`;

exports[`UI count basics 1`] = `
">> >> 1
>> 0
>> >> 2
>> >> "
`;

exports[`UI delete 1`] = `
">> >> 1234
>> >> NULL
>> >> "
`;

exports[`UI invalid commands 1`] = `
">> dont know what to do with hello
>> Invalid command HELLO
>> Invalid GET
>> Invalid SET
>> Invalid SET
>> Invalid DELETE
>> >> "
`;
